src/tokenizer/shell.cpp

```diff
@@ -0,0 +1,526 @@
+            } else if (current_shell_scope_ == shell_scope::process_substitution) {
+                add_token(shell_token_type::process_substitution_close);
+                pop_shell_scope();
+            } else if (current_shell_scope_ == shell_scope::subshell) {
+                add_token(shell_token_type::subshell_close);
+                pop_shell_scope();
+            } else {
+                add_token(shell_token_type::parenthesis_close);
+            }
+        } else if (c == '`') {
+            if (current_shell_scope_ == shell_scope::backtick_substitution) {
+                // End of the backtick command substitution, add a token for the
+                // final backtick and exit the scope
+                add_token(shell_token_type::backtick_substitution_close);
+
+                pop_shell_scope();
+            } else {
+                // Backtick substitution, add a token for the first backtick and
+                // open a new scope
+                add_token(shell_token_type::backtick_substitution_open);
+                push_shell_scope(shell_scope::backtick_substitution);
+            }
+        } else if (c == '(') {
+            if (!tokens_.empty() && current_token_type() == shell_token_type::equal) {
+                // Array
+                tokenize_delimited_token(")", shell_token_type::field);
+            } else if (next() == '(') {
+                // Arithmetic expansions
+                tokenize_delimited_token("))", shell_token_type::arithmetic_expansion);
+            } else if (should_expect_subprocess()) {
+                add_token(shell_token_type::subshell_open);
+                push_shell_scope(shell_scope::subshell);
+            } else {
+                add_token(shell_token_type::parenthesis_open);
+            }
+        } else if (c == '=') {
+            add_token(shell_token_type::equal);
+        } else if (c == '\n' || c == ';' || c == '!') {
+            add_token(shell_token_type::control);
+        } else if (c == '|') {
+            add_token(shell_token_type::control, next() == '|' ? 2 : 1);
+        } else if (c == '&') {
+            auto n = next();
+            if (n == '>') {
+                tokenize_redirection();
+            } else if (n == '&') {
+                add_token(shell_token_type::control, 2);
+            } else {
+                add_token(shell_token_type::control);
+            }
+        } else if (c == '{') {
+            auto n = next();
+            if (n == ' ') {
+                add_token(shell_token_type::compound_command_open);
+                // Swallow the whitespace
+                advance();
+                push_shell_scope(shell_scope::compound_command);
+            } else {
+                add_token(shell_token_type::curly_brace_open);
+            }
+        } else if (c == '}') {
+            if (current_shell_scope_ == shell_scope::compound_command) {
+                add_token(shell_token_type::compound_command_close);
```

Comment:
for close, `}` is not enough:

```
$ { echo 1 }; }
1 }
```

only the 2nd is effective

Reply:
This has been fixed now.